### PR TITLE
JAMES-2283 Subscribe to system mailboxes upon JMAP provisioning

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java [hunk 1]
```diff
@@ -96,7 +97,20 @@ public void provisionMailboxesShouldNotDuplicateMailboxByName() throws Exception
             .body(NAME, equalTo("mailboxes"))
             .body(ARGUMENTS + ".list", hasSize(5))
             .body(ARGUMENTS + ".list.name", hasItems(DefaultMailboxes.DEFAULT_MAILBOXES.toArray()));
+    }
+
+    @Test
+    public void provisionMailboxesShouldRegisterThem() throws Exception {
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java]: s/Register/SubscribeTo/ ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterTest.java [hunk 2]
```diff
@@ -62,6 +65,14 @@ public void createMailboxesIfNeededShouldCreateSystemMailboxes() throws Exceptio
                 .collect(Guavate.toImmutableList()));
     }
 
+    @Test
+    public void createMailboxesIfNeededShouldSubscribeMailboxes() throws Exception {
+        testee.createMailboxesIfNeeded(session);
+
+        assertThat(subscriptionManager.subscriptions(session))
+            .containsOnlyElementsOf(DefaultMailboxes.DEFAULT_MAILBOXES);
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterTest.java]: why `Only` ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterThreadTest.java [hunk 3]
```diff
@@ -84,7 +89,34 @@ public void testConcurrentAccessToFilterShouldNotThrow() {
         AnnotatedTestRunner runner = new AnnotatedTestRunner();
         runner.runTests(this.getClass(), DefaultMailboxesProvisioningFilter.class);
     }
-    
+
+    private static class FakeSubscriptionManager implements SubscriptionManager {
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterThreadTest.java]: please use a mock instead

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java [hunk 4]
```diff
@@ -97,7 +98,20 @@ public void provisionMailboxesShouldNotDuplicateMailboxByName() throws Exception
...
+        String token = HttpJmapAuthentication.authenticateJamesUser(baseUri(), USER, PASSWORD).serialize();
+
+        with()
+            .header("Authorization", token)
+            .body("[[\"getMailboxes\", {}, \"#0\"]]")
+            .post("/jmap");
 
+        assertThat(jmapServer.getProbe(MailboxProbeImpl.class)
+            .listSubscriptions(USER))
+            .containsOnlyElementsOf(DefaultMailboxes.DEFAULT_MAILBOXES);
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java]: why `Only` ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java [hunk 5]
```diff
@@ -387,7 +388,14 @@ public void subscriptionUserShouldBeChangedWhenCreateThenUpdateMailboxNameWithJM
...
 
-        assertThat(mailboxProbe.listSubscriptions(username)).containsOnly("mySecondBox");
+        List<String> expectedMailboxes = ImmutableList.<String> builder()
+            .add(DefaultMailboxes.OUTBOX)
+            .add(DefaultMailboxes.SENT)
+            .add(DefaultMailboxes.TRASH)
+            .add(DefaultMailboxes.DRAFTS)
+            .add("mySecondBox")
+            .build();
+        assertThat(mailboxProbe.listSubscriptions(username)).containsOnlyElementsOf(expectedMailboxes);
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: you probably don't need to build the list in the first place. Why don't you use containsOnly(...) instead ?

Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: why `Only` here too ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java [hunk 6]
```diff
@@ -411,7 +419,13 @@ public void subscriptionUserShouldBeDeletedWhenDestroyMailbox() throws Exception
...
             .statusCode(200);
 
-        assertThat(mailboxProbe.listSubscriptions(username)).isEmpty();
+        List<String> expectedMailboxes = ImmutableList.<String> builder()
+                .add(DefaultMailboxes.OUTBOX)
+                .add(DefaultMailboxes.SENT)
+                .add(DefaultMailboxes.TRASH)
+                .add(DefaultMailboxes.DRAFTS)
+                .build();
+        assertThat(mailboxProbe.listSubscriptions(username)).containsOnlyElementsOf(expectedMailboxes);
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: same comments holds for this one and next ones

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterThreadTest.java (hunk 3 above)
Comment (chibenwa) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterThreadTest.java]: You can not! It's due to that concurrent testing framework... Sorry, it do not combine well

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java (hunk 4 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java]: Order of elements? We care about the elements in there, not their order, hence it is a containOnly.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 5 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: Same answer. Do we care about the order?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterTest.java (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterTest.java]: Yet again, do we care about the order?

**Comment by chibenwa** [comment]
test this please (not found)

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterThreadTest.java (hunk 3 above)
Comment (mbaechler) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterThreadTest.java]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java (hunk 4 above)
Comment (mbaechler) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java]: You got me wrong ; we don't need order nor strict equality so `contains` should be enough. My question was : why would we need to check there's nothing else returned ?

**Comment by chibenwa** [comment]
```
[267ca6df09474b12a9ea7819d1ab51399e64b6ca] [ERROR] Failures: 
[267ca6df09474b12a9ea7819d1ab51399e64b6ca] [ERROR]   CassandraSetMailboxesMethodTest>SetMailboxesMethodTest.subscriptionUserShouldBeDeletedWhenCreateThenDestroyMailboxWithJMAP:473 
[267ca6df09474b12a9ea7819d1ab51399e64b6ca] Expecting:
[267ca6df09474b12a9ea7819d1ab51399e64b6ca]   <["Outbox", "Trash", "Drafts", "Sent"]>
[267ca6df09474b12a9ea7819d1ab51399e64b6ca] to contain only:
[267ca6df09474b12a9ea7819d1ab51399e64b6ca]   <["Outbox", "Sent", "Trash", "Drafts", "mySecondBox"]>
[267ca6df09474b12a9ea7819d1ab51399e64b6ca] but could not find the following elements:
[267ca6df09474b12a9ea7819d1ab51399e64b6ca]   <["mySecondBox"]>
```

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java (hunk 4 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java]: Hint: that is what I first wrote ;-)

Comment (mbaechler) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/ProvisioningTest.java]: it's a general principle : don't assert more than the test description assumes

**Comment by chibenwa** [comment]
The rationals: explain what I am doing. I can't pass custom messages in fixups.

**Comment by aduprat** [comment]
You can push a comment on the PR just after pushing.

**Comment by rouazana** [comment]
merged

**Pull request closed without merging**